### PR TITLE
Add more tests to the autograd wrt view and inplace

### PR DESCRIPTION
**Pull request by albanD**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31147 Add more tests to the autograd wrt view and inplace**

The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

Differential Revision: [D19301082](https://our.internmc.facebook.com/intern/diff/D19301082)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 6258c79:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6258c79855fa97525a631961d435e416cfac27cc).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| CircleCI <sup>pytorch_linux_xenial_py3_clang5_mobile_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: test/test_autograd.py [hunk 1]
```diff
@@ -3333,6 +3333,318 @@ def test_checkpointing(self):
         mean_combined = torch.stack(feat_combined).mean()
         mean_combined.backward()
 
+    def test_autograd_views_codegen(self):
+        # This is not necessarily the absolute correct behavior, but this is the current
+        # one. This test is here to make sure that any change to this behavior is detected
+        # and not silent.
```
Comment (ezyang) [inline test/test_autograd.py]: Maybe, explain what acceptable other behaviors are? I.e. what are the degrees of freedom in the test?

File: test/test_autograd.py [hunk 2]
```diff
@@ -3333,6 +3333,318 @@ def test_checkpointing(self):
...
+            out = get_unbind()
+            maybe_check_raise(lambda: out[0].add_(1), should_raise_tuple[1])
+            maybe_check_raise(lambda: out[1].add_(1), should_raise_tuple[2])
+
+
+        # Args: grad_mode, requires_grad, is_view, should_raise
+        # should_raise contains None if it should not raise
+        # should_raise contains a string of the error if it should raise
+        # The 3 elements are for view_as, first output of unbind and second output of unbind
+        run_test(True, False, True, (None, None, None))
```
Comment (ezyang) [inline test/test_autograd.py]: These tests would be easier to read with kwargs, IMO.

File: test/test_autograd.py (hunk 1 above)
Comment (albanD) [reply test/test_autograd.py]: Changing any of these test will be a BC-breaking change. All these tests use only public api so users could be using it.

File: test/test_autograd.py [hunk 3]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
         mean_combined = torch.stack(feat_combined).mean()
         mean_combined.backward()
 
+    def test_autograd_views_codegen(self):
+        # This is not necessarily the absolute correct behavior, but this is the current
+        # one. This test is here to make sure that any change to this behavior is detected
+        # and not silent. The TODOs below mark the places with unexpected behavior.
+        # Note that any change in these test will be BC-breaking and should be done carefully.
+
+        # This test checks the behavior that sample codegen functions (view_as and unbind)
```
Comment (gchanan) [inline test/test_autograd.py]: I'm not sure what "sample codegen functions" means?

File: test/test_autograd.py [hunk 4]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+                    return inp.view_as(inp)
+
+
+            def get_unbind():
+                inp = torch.rand(2, requires_grad=requires_grad).clone()
+                with torch.set_grad_enabled(grad_mode):
+                    return inp.unbind()
+
+            def maybe_check_raise(fn, should_raise):
+                if should_raise is not None:
```
Comment (gchanan) [inline test/test_autograd.py]: super nit: prefer "if should_raise:" -- it would be surprising if someone passed False to should_raise and it was interpreted as True.

File: test/test_autograd.py [hunk 5]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+            maybe_check_raise(lambda: out[0].add_(1), should_raise_tuple[1])
+            maybe_check_raise(lambda: out[1].add_(1), should_raise_tuple[2])
+
+
+        # should_raise contains None if it should not raise
+        # should_raise contains a string of the error if it should raise
+        # The 3 elements are for view_as, first output of unbind and second output of unbind
+        run_test(grad_mode=True, requires_grad=False, is_view=True,
+                 should_raise_tuple=(None, None, None))
+        # TODO: Second should_raise should not be None below
```
Comment (gchanan) [inline test/test_autograd.py]: interesting -- the second return of unbind is different than the first?

File: test/test_autograd.py [hunk 6]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+        run_test(grad_mode=True, requires_grad=True, is_view=True,
+                 should_raise_tuple=(None, None, "INTERNAL ASSERT FAILED"))
+        # TODO: views require gradients when created in no_grad mode but their grad_fn is not populated
+        run_test(grad_mode=False, requires_grad=True, is_view=True,
+                 should_raise_tuple=("leaf Variable that requires grad",
```
Comment (gchanan) [inline test/test_autograd.py]: nit: I'd just put the full error message -- I spent awhile trying to guess what the actual error was.

File: test/test_autograd.py (hunk 5 above)
Comment (albanD) [reply test/test_autograd.py]: The view is only checked at runtime and the first output being a view is allowed.
The second output is not though.

But these cases should not even exist: we should never contruct such Node. This is (should be) enforced when constructing Nodes for an autograd.Function. But the codegen does it by hand so no other check happen at construction time.

File: test/test_autograd.py (hunk 6 above)
Comment (albanD) [reply test/test_autograd.py]: I wanted to reduce line size. But I'll get a common string with the full error message.

File: test/test_autograd.py (hunk 3 above)
Comment (albanD) [reply test/test_autograd.py]: Subset of functions from the codegen selected by hand. I'll say that explicitly.

File: test/test_autograd.py (hunk 4 above)
Comment (albanD) [reply test/test_autograd.py]: I prefer None vs string to False vs string.
I'll add an assert that it is either None or a string to make sure this does not happen.

File: test/test_autograd.py [hunk 7]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+
+
+        # should_raise contains None if it should not raise
+        # should_raise contains a string of the error if it should raise
+        # The 3 elements are for view_as, first output of unbind and second output of unbind
+        run_test(grad_mode=True, requires_grad=False, is_view=True,
+                 should_raise_tuple=(None, None, None))
+        # TODO: Second should_raise should not be None below
+        run_test(grad_mode=True, requires_grad=True, is_view=True,
+                 should_raise_tuple=(None, None, "diff_view_meta->output_nr_ == 0 INTERNAL ASSERT FAILED"))
```
Comment (ezyang) [inline test/test_autograd.py]: The contract of INTERNAL ASSERT is that it should be impossible to trigger it from userland (aka Python). So this is just a straight up bug; either the assert is incorrectly classified, or we're doing something wrong here

File: test/test_autograd.py [hunk 8]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+            # Are they differentiable views?
+            out = get_view_as()
+            self.assertTrue(out._is_view() == is_view)
+
+            out = get_unbind()
+            self.assertTrue(out[0]._is_view() == is_view)
+            self.assertTrue(out[1]._is_view() == is_view)
+
+            # Are inplace allowed?
+            out = get_view_as()
```
Comment (ezyang) [inline test/test_autograd.py]: I don't understand, why do you need to allocate fresh variables for the inplace test? The code here would be a lot more direct if you just inlined the `view_as` and `get_unbind` implementations; they seem to be factored into functions so you can call them twice but the second call seems completely unnecessary.

File: test/test_autograd.py [hunk 9]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+    def test_autograd_views_codegen(self):
+        # This is not necessarily the absolute correct behavior, but this is the current
+        # one. This test is here to make sure that any change to this behavior is detected
+        # and not silent. The TODOs below mark the places with unexpected behavior.
+        # Note that any change in these test will be BC-breaking and should be done carefully.
+
+        # This test checks the behavior of two codegen functions (view_as and unbind)
+        # with respect to view tracking and inplace operation on the output.
+
+        def run_test(grad_mode, requires_grad, is_view, should_raise_tuple):
```
Comment (ezyang) [inline test/test_autograd.py]: The ordering of the arguments here is a little backward. First you allocate the variable (using `requires_grad`), then you run the operation with grad enabled or disabled (`grad_mode`). But in the argument list here they're reverse order

Comment (ezyang) [reply test/test_autograd.py]: In every usage of this function, `is_view` is true. I don't think we're ever going to make `view_as` or `unbind` return not views? So why have them be arguments to this function?

File: test/test_autograd.py [hunk 10]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+
+        # I) This checks the autograd.Function behavior when we return one or multiple outputs
+        # while one of these is an input, a view of an input or of a temporary tensor.
+        # II) This checks that multiples views in the forward are properly traced and how they
+        # behave with respect to inplace operations.
+
+
+        # I) Simple views
+        bw_called = [0]
+        ga_nz = [False]
```
Comment (ezyang) [inline test/test_autograd.py]: I think it would be very helpful for readers if it was explained what these indicator variables represent

File: test/test_autograd.py [hunk 11]
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
...
+                        # TODO: The custom backward is not called in this case
+                        expected_called = 0
+                    if fn_id == "two_output" and inplace:
+                        # TODO: The backward only sees part of the gradient as the other part
+                        # is computed with a AsStridedBackward
+                        expected_ga_nz = False
+                    self.assertTrue(bw_called[0] == expected_called)
+                    self.assertTrue(ga_nz[0] == expected_ga_nz)
+
+        # II) Complex views
```
Comment (ezyang) [inline test/test_autograd.py]: Split this into a separate test

**Review by ezyang — APPROVED** [review]

File: test/test_autograd.py (hunk 9 above)
Comment (albanD) [reply test/test_autograd.py]: Well I would argue that we should not return views in `torch.no_grad()` mode.
Because here we "autodiff" the view op even though it was in a no grad block.
See https://github.com/pytorch/pytorch/issues/26546 for reference of this problem.

File: test/test_autograd.py (hunk 7 above)
Comment (albanD) [reply test/test_autograd.py]: Yes, both the 2nd and 3rd here should raise an error stating that we're doing an inplace on the output of a function where it's not allowed.

File: test/test_autograd.py (hunk 9 above)
Comment (ezyang) [reply test/test_autograd.py]: oooooh this is giving me flashbacks to another issue. This one: https://github.com/pytorch/pytorch/pull/12502 (and also #11390)

File: test/test_autograd.py (hunk 8 above)
Comment (albanD) [reply test/test_autograd.py]: Yes I'll clean this up.
